### ISSUE TITLE
Object values are not assignable when wrapped in a Partial

### ISSUE THREAD

**Issue by rraval** [Microsoft/TypeScript#26266]
**TypeScript Version:**  Version 3.1.0-dev.20180807

**Search Terms:** keyof mapped partial

**Code**

Compiled with `--strictNullChecks`:

```ts
interface Value<T> {
    value: T;
}

interface Builder<T> {
    buildValue(): Value<T>;
}

type BuilderMap<Keys extends PropertyKey> = {
    [K in Keys]: Builder<any>;
}

type ValueMap<Keys extends PropertyKey> = {
    [K in Keys]: Value<any>;
};

function makeFooMap<Map extends BuilderMap<"foo">>(builderMap: Map) {
    const valueMap: Partial<ValueMap<keyof Map>> = {};
    valueMap["foo"] = builderMap["foo"].buildValue();
}
```

**Expected behavior:**

The code compiles without errors.

**Actual behavior:**

The compiler dislikes the `valueMap["foo"]` assignment:

```
bug.ts:19:5 - error TS2322: Type 'Value<any>' is not assignable to type 'ValueMap<keyof Map>["foo"]'.

19     valueMap["foo"] = builderMap["foo"].buildValue();
       ~~~~~~~~~~~~~~~
```

Tested versions:

- 2.9.2: no error
- 3.0.1: reports error
- 3.1.0-dev.20180807: reports error

Changing the code to remove the `Partial` and using a type assertion seems to work:

```ts
function makeFooMap<Map extends BuilderMap<"foo">>(builderMap: Map) {
    const valueMap: ValueMap<keyof Map> = {} as ValueMap<keyof Map>;
    valueMap["foo"] = builderMap["foo"].buildValue(); // no error anymore
}
```

Getting rid of the `keyof` and explicitly providing a list of keys also works:

```ts
function makeFooMap<Map extends BuilderMap<"foo">>(builderMap: Map) {
    const valueMap: Partial<ValueMap<"foo">> = {};
    valueMap["foo"] = builderMap["foo"].buildValue(); // no error anymore
}
```

**Playground Link:** [Playground Link](https://www.typescriptlang.org/play/#src=interface%20Value%3CT%3E%20%7B%0A%20%20%20%20value%3A%20T%3B%0A%7D%0A%0Ainterface%20Builder%3CT%3E%20%7B%0A%20%20%20%20buildValue()%3A%20Value%3CT%3E%3B%0A%7D%0A%0Atype%20BuilderMap%3CKeys%20extends%20PropertyKey%3E%20%3D%20%7B%0A%20%20%20%20%5BK%20in%20Keys%5D%3A%20Builder%3Cany%3E%3B%0A%7D%0A%0Atype%20ValueMap%3CKeys%20extends%20PropertyKey%3E%20%3D%20%7B%0A%20%20%20%20%5BK%20in%20Keys%5D%3A%20Value%3Cany%3E%3B%0A%7D%3B%0A%0Afunction%20makeFooMap%3CMap%20extends%20BuilderMap%3C%22foo%22%3E%3E(builderMap%3A%20Map)%20%7B%0A%20%20%20%20const%20valueMap%3A%20Partial%3CValueMap%3Ckeyof%20Map%3E%3E%20%3D%20%7B%7D%3B%0A%20%20%20%20valueMap%5B%22foo%22%5D%20%3D%20builderMap%5B%22foo%22%5D.buildValue()%3B%0A%7D)

**Related Issues:** #25010 maybe?

**Comment by andy-ms** [comment]
```ts
interface Value { x: number };

type Values<Keys extends PropertyKey> = { [K in Keys]: Value; }

function f0<T extends { "foo": number }>(_: T, value: Value): Partial<Values<keyof T>> {
    const valueMap: Partial<Values<keyof T>> = {};
    const x: Value | undefined = valueMap["foo"]; // No error
    valueMap["foo"] = value; // Error?
    return valueMap;
}

function f1<K extends string>(x: Values<K>): Partial<Values<K>> {
    return x; // Error?
}
```

No error if using `type Value = number;`. No error if you remove ` extends { "foo": number }`.
I would expect `Value` to be assignable to `valueMap["foo"]` which is of type `Value | undefined` (regardless of the type of `keyof T`).
I would also expect `Values<K>` to be assignable to `Partial<Values<K>>` regardless of the value of `K`.

**Comment by rraval** [comment]
@andy-ms I'm not sure if this is related or if this should be a separate bug report, but I'm seeing some weird behaviour with `Pick` as well:

```ts
type Values<Keys extends PropertyKey> = {
    [K in Keys]: number;
};

function f0<Keys extends PropertyKey>(attrs: Values<Keys>) {
    for (const key in attrs) {
        const value: number = attrs[key]; // No error
    }
}

function f1<Keys extends PropertyKey>(attrs: Pick<Values<Keys>, Keys>) {
    for (const key in attrs) {
        const value: number = attrs[key]; // Error
    }
}
```

The error in question is:

```
bug.ts:13:15 - error TS2322: Type 'Values<Keys>[Extract<Keys, string>]' is not assignable to type 'number'.

13         const value: number = attrs[key]; // Error
                 ~~~~~
```

In my head, wrapping an object in `Pick` shouldn't affect my ability to read out values from the object, but that may just be my mistaken understanding.

**Comment by ahejlsberg** [comment]
This was fixed by #26698. All of the above examples now type check with no errors.
